detectors/gcp/utils_test.go

```diff
@@ -14,62 +14,87 @@
 
 package gcp
 
-func NewTestDetector(metadata *FakeMetadataProvider, os *FakeOSProvider) *Detector {
-	return &Detector{metadata: metadata, os: os}
+import (
+	"fmt"
+	"strings"
+
+	"cloud.google.com/go/compute/metadata"
+)
+
+func NewTestDetector(metadataProvider *FakeMetadataProvider, os *FakeOSProvider) *Detector {
+	return &Detector{metadata: metadataProvider, os: os}
 }
 
 type FakeMetadataProvider struct {
-	Err                  error
-	Attributes           map[string]string
-	InstanceAttributes   map[string]string
-	FakeInstanceID       string
-	FakeInstanceName     string
-	FakeInstanceHostname string
-	FakeZone             string
-	Project              string
+	Err      error
+	Metadata map[string]string
 }
 
-func (f *FakeMetadataProvider) ProjectID() (string, error) {
+func (f *FakeMetadataProvider) Get(s string) (string, error) {
 	if f.Err != nil {
 		return "", f.Err
 	}
-	return f.Project, nil
-}
-func (f *FakeMetadataProvider) InstanceID() (string, error) {
-	if f.Err != nil {
-		return "", f.Err
+	value, ok := f.Metadata[s]
+	if ok {
+		return value, nil
 	}
-	return f.FakeInstanceID, nil
+	return "", metadata.NotDefinedError(s)
 }
-func (f *FakeMetadataProvider) Get(s string) (string, error) {
-	if f.Err != nil {
-		return "", f.Err
-	}
-	return f.Attributes[s], nil
+
+// These keys are all documented at https://cloud.google.com/compute/docs/metadata/predefined-metadata-keys
+func (f *FakeMetadataProvider) ProjectID() (string, error) {
+	return f.Get("project/project-id")
+}
+func (f *FakeMetadataProvider) InstanceID() (string, error) {
+	return f.Get("instance/id")
 }
 func (f *FakeMetadataProvider) InstanceName() (string, error) {
-	if f.Err != nil {
-		return "", f.Err
-	}
-	return f.FakeInstanceName, nil
+	return f.Get("instance/name")
 }
 func (f *FakeMetadataProvider) Hostname() (string, error) {
-	if f.Err != nil {
-		return "", f.Err
-	}
-	return f.FakeInstanceHostname, nil
+	return f.Get("instance/hostname")
 }
 func (f *FakeMetadataProvider) Zone() (string, error) {
-	if f.Err != nil {
-		return "", f.Err
+	value, err := f.Get("instance/zone")
+	if err != nil {
+		return value, err
 	}
-	return f.FakeZone, nil
+	// https://github.com/googleapis/google-cloud-go/blob/35d1f813f755c88d6869d993796c5ddf8ac3eea3/compute/metadata/metadata.go#L663
+	return value[strings.LastIndex(value, "/")+1:], nil
```

Comment:
I'm not sure this is helpful for testing our library, even if it is closer to how the metadata library works.

Reply:
Yeah, I was torn on whether to do this or not, but I came down on the side of wanting `Get("instance/zone")` and `Zone()` to always return matching values, which means copying this logic.

Now that I've done this simplification we could also get rid of `FakeMetadatProvider` entirely if we want and just use the upstream `metadata.Client` with a mock HTTP handler. Then we'd use upstream's split logic here. WDYT?

Reply:
sounds good.  lets do it